controller/kubernetes_controller.go

```diff
@@ -0,0 +1,382 @@
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (kc *KubernetesController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+
+	err := kc.syncKubernetesStatus(key.(string))
+	kc.handleErr(err, key)
+
+	return true
+}
+
+func (kc *KubernetesController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("Error syncing Longhorn volume kubernetes status %v: %v", key, err)
+		kc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Persistent Volume %v out of the queue: %v", key, err)
+	kc.queue.Forget(key)
+}
+
+func (kc *KubernetesController) syncKubernetesStatus(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "longhorn kubernetes-controller: fail to sync %v", key)
+	}()
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	if err = kc.cleanupForPVDeletion(name); err != nil {
+		return err
+	}
+
+	pv, err := kc.kubeClient.CoreV1().PersistentVolumes().Get(name, metav1.GetOptions{})
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "Error getting Persistent Volume %s", name)
+	}
+	volumeName := kc.validatePersistentVolume(pv)
+	if volumeName == "" {
+		return nil
+	}
+
+	volume, err := kc.ds.GetVolume(volumeName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			logrus.Infof("Longhorn volume %v has been deleted", volumeName)
+			return nil
+		}
+		return err
+	}
+
+	// existing volume may be used/reused by pv
+	if volume.Status.KubernetesStatus.PVName != name {
+		volume.Status.KubernetesStatus = types.KubernetesStatus{}
+		kc.eventRecorder.Eventf(volume, v1.EventTypeNormal, EventReasonStart, "Longhorn volume %v is being used by Persistent Volume %v", volume.Name, name)
+	}
+	ks := &volume.Status.KubernetesStatus
+
+	ks.PVName = name
+	if pv.Spec.ClaimRef != nil {
+		if pv.Status.Phase == v1.VolumeBound {
+			// set for bounded PVC
+			ks.PVCName = pv.Spec.ClaimRef.Name
+			ks.Namespace = pv.Spec.ClaimRef.Namespace
+			ks.LastPVCRefAt = ""
+		} else {
+			// PVC is no longer bound with PV. indicating history data by setting <LastPVCRefAt>
+			if ks.PVStatus == string(v1.VolumeBound) && ks.LastPVCRefAt == "" {
+				ks.LastPVCRefAt = util.Now()
+			}
+		}
+	} else {
+		if ks.LastPVCRefAt == "" {
+			if pv.Status.Phase == v1.VolumeBound {
+				return fmt.Errorf("BUG: PV with Bound phase has no ClaimRef")
+			}
+			// The associated PVC is removed from the PV ClaimRef
+			if ks.PVCName != "" {
+				ks.LastPVCRefAt = util.Now()
+			}
+		}
+	}
+	ks.PVStatus = string(pv.Status.Phase)
+
+	var pod *v1.Pod
+	pod, err = kc.getAssociatedPod(ks.Namespace, ks.PVCName)
```

Comment:
Just do `pod, err :=`

`:=` works as long as at least one variable on the left is new.